src/kademlia/dht_storage.cpp

```diff
@@ -0,0 +1,355 @@
+		, tcp::endpoint const& endp
+		, std::string const& name, bool seed)
+	{
+		table_t::iterator ti = m_map.find(info_hash);
+		torrent_entry* v;
+		if (ti == m_map.end())
+		{
+			// we don't have this torrent, add it
+			// do we need to remove another one first?
+			if (!m_map.empty() && int(m_map.size()) >= m_settings.max_torrents)
+			{
+				// we need to remove some. Remove the ones with the
+				// fewest peers
+				int num_peers = m_map.begin()->second.peers.size();
+				table_t::iterator candidate = m_map.begin();
+				for (table_t::iterator i = m_map.begin()
+					 , end(m_map.end()); i != end; ++i)
+				{
+					if (int(i->second.peers.size()) > num_peers) continue;
+					if (i->first == info_hash) continue;
+					num_peers = i->second.peers.size();
+					candidate = i;
+				}
+				m_map.erase(candidate);
+				m_counters.inc_stats_counter(counters::dht_torrents, -1);
+			}
+			m_counters.inc_stats_counter(counters::dht_torrents);
+			v = &m_map[info_hash];
+		}
+		else
+		{
+			v = &ti->second;
+		}
+
+		// the peer announces a torrent name, and we don't have a name
+		// for this torrent. Store it.
+		if (!name.empty() && v->name.empty())
+		{
+			std::string tname = name;
+			if (tname.size() > 50) tname.resize(50);
+			v->name = tname;
+		}
+
+		peer_entry peer;
+		peer.addr = endp;
+		peer.added = aux::time_now();
+		peer.seed = seed;
+		std::set<peer_entry>::iterator i = v->peers.find(peer);
+		if (i != v->peers.end()) v->peers.erase(i++);
+		v->peers.insert(i, peer);
+	}
+
+	bool dht_default_storage::get_immutable_item(sha1_hash const& target
+		, dht_immutable_item& item) const
+	{
+		dht_immutable_table_t::const_iterator i = m_immutable_table.find(target);
+		if (i == m_immutable_table.end()) return false;
+
+		item = i->second.item;
+		return true;
+	}
+
+	void dht_default_storage::put_immutable_item(sha1_hash const& target, dht_immutable_item const& to_add, address addr)
+	{
+		dht_immutable_table_t::iterator i = m_immutable_table.find(target);
+		if (i == m_immutable_table.end())
+		{
+			// make sure we don't add too many items
+			if (int(m_immutable_table.size()) >= m_settings.max_dht_items)
+			{
+				// delete the least important one (i.e. the one
+				// the fewest peers are announcing, and farthest
+				// from our node ID)
+				dht_immutable_table_t::iterator j = std::min_element(m_immutable_table.begin()
+						, m_immutable_table.end()
+						, immutable_item_comparator(m_id));
+
+				TORRENT_ASSERT(j != m_immutable_table.end());
+				free(j->second.item.value);
+				m_immutable_table.erase(j);
+				m_counters.inc_stats_counter(counters::dht_immutable_data, -1);
+			}
+
+			boost::tie(i, boost::tuples::ignore) = m_immutable_table.insert(
+					std::make_pair(target, to_add));
+			m_counters.inc_stats_counter(counters::dht_immutable_data);
+		}
+
+		touch_item(&i->second.popularity, addr);
+	}
+
+	bool dht_default_storage::get_mutable_item(sha1_hash const& target
```

Comment:
this may be a bit premature for v1 of this patch, but do you have any ideas of how the read operations could be made asynchronous? I'm worried about disk I/O blocking the main thread of libtorrent when using sqlite3 as the backend.